### ISSUE TITLE
Add collection of cycles and instructions

### ISSUE THREAD

**Issue by yonch** [perfpod/memory-collector#23]
Cycles per instruction (CPI) shows how efficient the CPU is in executing its instructions. When the CPU has to wait for memory reads because of cache contention and/or DRAM contention, then the CPU experiences _stall cycles_: cycles where it is just waiting without doing useful work. By examining the ratio of cycles to instructions (the CPI), it is possible to detect periods of high contention (which also mean low application efficiency).

In this issue, we add a preliminary collection of cycles per instruction: monitoring the current process's cycles and instructions. 

What this helps us learn (in addition to the code):
- Gives us experience with perf collection
- Are there pre-requisites to reading these counters that we are not aware of? (like, do we need to mount a file-system or run the operating system with a specific parameters)
- Are open-source packages sufficient for gathering these metrics (this will show we don't have to build our own from scratch)
- Does the package output match the canonical Linux tools (`perf`).

The `perf` linux subsystem allows collection of multiple counters. `perf` command-line tool interacts with the command line system, and [`perf stat`](https://man7.org/linux/man-pages/man1/perf-stat.1.html) collects the counters we're looking for. golang's [elastic/go-perf](https://github.com/elastic/go-perf/tree/master) seems like a good fit for golang collection.

Suggested steps:
- [ ] Start a binary (in the `cmd/` directory).
- [ ] Add a function that reads the cycles and instructions in the current process in a 10 second interval and returns the read values. Call it from `main()` and sanity check.
- [ ] Implement a function that runs the equivalent `perf stat` command, and returns its output. Add it to the start of `main()` and check sanity.
- [ ] Add a function that burns some cycles with an arbitrary computation. (note: it needs to be written such that the compiler doesn't optimize it away). sanity test in main.
- [ ] Now run the three functions in parallel in main (using goroutines, e.g., `go read_perf_counters();`, `go run_perf_stat_command()`, and `go compute_heavy_test_workload()`), and print both results, manually verify they return a similar result.
- [ ] Add this as a test, and verify the returned values are similar (with some fudge factor for the values to diverge since the measurement intervals of the two functions might not be exactly the same).
- [ ] Add a Github Actions workflow that runs the test.

Can split off steps if this is too big.

# Using elastic/go-perf 
These are pointers to get started, not an exhaustive. Please reach out on the issue to discuss if there are gaps that make this difficult!

Here is the [documentation](https://pkg.go.dev/github.com/elastic/go-perf#section-readme).

In the [Counting Events](https://pkg.go.dev/github.com/elastic/go-perf#hdr-Counting_events) section, the first snippet shows how to gather Cycles and Instruction for a specific Process ID. [`testIPC()` in `count_test.go`](https://github.com/elastic/go-perf/blob/30bec95324b80929a97cb30b61bf408715c02454/count_test.go#L34) is a fuller implementation.

**Comment by yonch** [comment]
@tverghis you were looking for issues, this is the first of the batch

**Comment by tverghis** [comment]
Thanks for the ping, and the detailed write-up! I'll give this a shot.

**Comment by tverghis** [comment]
I've started the implementation here, in my fork: https://github.com/tverghis/memory-collector/blob/cpi_count/cmd/cpi-count/main.go

This works as expected - here is some sample output:
```
[23:17] borzoi:cpi-count (cpi_count)| sudo ./cpi-count
2024/12/19 23:17:05 Ran for 0ms
2024/12/19 23:17:05 Cycles: 109189, Instrs: 41833, CPI: 2.610116
2024/12/19 23:17:05 Finished running, closing event group.
```

However, this doesn't (yet?) necessarily meet the ask of this checklist item:
>Add a function that reads the cycles and instructions in the current process in a 10 second interval and returns the read values

Namely because [`Event.MeasureGroup`](https://pkg.go.dev/github.com/elastic/go-perf#Event.MeasureGroup) specifically instruments the function passed to it; it doesn't seem to just measure all events for the current process.

Browsing the examples in elastic/go-perf, I think what we probably want instead is probably something closer to [this example](https://github.com/elastic/go-perf/blob/30bec95324b80929a97cb30b61bf408715c02454/perf_example_test.go#L111)?

**Comment by yonch** [comment]
It seems you're right @tverghis! The documentation for [`Open()`](https://pkg.go.dev/github.com/elastic/go-perf#Open) (which is used in the implementation of Group.Open) appears to talk about threads, not processes:

> if pid == CallingThread and cpu == AnyCPU, the event measures the calling thread on any CPU

This issue is supposed to get an initial implementation working and verified with perf, and the follow up would be to monitor containers via [`OpenCGroup`](https://pkg.go.dev/github.com/elastic/go-perf#OpenCGroup), so I think we should probably minimize effort of trying to get process measurements, instead we can monitor threads for now and will then move to groups... 

If that makes sense to you, we might write all of the functionality in the issue into one function and just monitor that function.

I mean that function would:
- Start the go-perf monitor
- Start the `perf` command line in the background, monitoring the current thread
- Perform the example workload
- Collect the results from the two monitors and report

**Comment by tverghis** [comment]
I got instrumentation of the workload working with `perf stat` as well, and the results look promising! The delta between using `elastic/go-perf` and `perf stat` seems to be in a fairly small, acceptable range, at least for the simple workload I was instrumenting.

https://github.com/tverghis/memory-collector/commit/cc01680e812f78ec714e59eec1b7f7ba5cc0d581

```
[00:37] borzoi:cpi-count (cpi_count *)| sudo ./cpi-count
2024/12/21 00:37:43 Current PID: 53403
2024/12/21 00:37:43 Started perf cmd
2024/12/21 00:37:43 Sum is 49999995000000.000000
2024/12/21 00:37:43 Ran for 87ms
---> 2024/12/21 00:37:43 GoPerf Cycles: 121501212, GoPerf Instrs: 80612232, GoPerf CPI: 1.507230
---> 2024/12/21 00:37:43 PerfCmd Cycles: 121993486, PerfCmd Instrs: 81000784, PerfCmd CPI: 1.506078
2024/12/21 00:37:43 Finished running, closing event group.
```
Next steps:
- Try instrumenting a heavier workload to see if the skew gets worse
- Write this as a test that we can run via automation to verify that the skew for each counter is within an acceptable range (1%? 5%?)

**Comment by yonch** [comment]
@tverghis doing some tidying, so the first few subtasks here can be checked off, right? Only the GitHub actions remaining?

For running the test, note that @niminim998 added CI that creates containers for the binaries in #29 and #30 .

**Comment by tverghis** [comment]
Yes, just running automated tests in GHA is missing at this point. Shall we close this issue and wrap up the work in #31?
